DR-2262 DR-2259 DR-2121 DR-2274 Azure Snapshot fixes

nmalfroy opened the pull request:
This PR addresses a couple of straggler issues that I noticed when prepping for the demo:
- Snapshots in Azure now create a DRS URI
- Added a way to read parquet files directly sans-Synapse for tests (and use that to verify that filerefs are properly converted to drs uris)
- Make sure we return the selfuri field in the DRS object (the Python SDK fails when we don't have this)
- Removed the gcpRegion field from the dataset creation request body
- MD5s for Azure files are now stored as hex strings
- Fixed an NPE where we were checking for Firestore info in Azure snapshots
- Made the invalid cloud specification an error class

Going through commit-by-commit should be easier

nmalfroy committed:
Make sure that MD5s are stored as hex values

...and not Base64 (as they are presented in the Azure portal)

nmalfroy committed:
No longer require the gcpRegion field when creating a dataset in Azure

nmalfroy committed:
Add an exception for invalid cloud platform specification

nmalfroy committed:
Make sure that firestore region check only happens for GCP snapshot creation

nmalfroy committed:
Fix NPE on when fetching an azure snapshot

nmalfroy committed:
Make sure that we return the selfUri field for DRS objects

nmalfroy committed:
Make sure that DRS IDs are created correctly when creating snapshots

nmalfroy committed:
Add step in integration test to validate parquet data

nmalfroy committed:
Make sure we bracket the column name in the drs calc

pshapiro4broad commented on src/main/java/bio/terra/common/SynapseColumn.java:
```diff
@@ -43,6 +46,10 @@ public SynapseColumn requiresJSONCast(boolean requiresJSONCast) {
   }
 
+  public boolean getIsFileType() {
+    return FILE_TYPES.contains(getType());
```
This check is done in a bunch of places, it would be nice to only write this code once. Unfortunately I don't see a good way to add it to the openapi generated enum, so it would have to live in a utils class somewhere.

nmalfroy replied:
Yeah, that's a good point.  I think we could maybe have datatype utils live in one place

pshapiro4broad replied:
Maybe this should be `isDrsType()` since it's both file and dir types, and what we're using it for is to determine if it needs to be converted to a DRS URL.

nmalfroy replied:
I don't want it to be drs specifically since we may be allowing generating different types of file uris.  At this point, it's really just that it's a file

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/DrsIdService.java:
```diff
@@ -43,23 +43,27 @@ private DrsId fromParts(String snapshotId, String fsObjectId) {
   }
 
-  public DrsId fromUri(String drsuri) {
+  public static DrsId fromUri(String drsuri) {
```
should probably be `drsUri`
```suggestion
  public static DrsId fromUri(String drsUri) {
```

nmalfroy replied:
Good point

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -157,6 +158,7 @@ public DRSObject lookupObjectByDrsId(
       } catch (IllegalArgumentException ex) {
         throw new InvalidDrsIdException("Invalid object id format '" + drsObjectId + "'", ex);
       } catch (SnapshotNotFoundException ex) {
+        logger.error("No snapshot found for DRS object id '" + drsObjectId + "'", ex);
```
In general, if you log as well as throw, you end up logging twice because the convention is to log at the "top" catch point (e.g. the catch point that doesn't re-throw). Is that not the case here?

nmalfroy replied:
I was struggling to find it in the logs (but that might also be that stackdriver was struggling).  I can remove

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java:
```diff
@@ -63,8 +71,14 @@
-          + "    OPENROWSET(\n"
+          + ") AS SELECT datarepo_row_id,<columns:{c|"
+          + "          <if(c.isFileType)>"
+          + "             'drs://<hostname>/v1_<snapshotId>_' + [<c.name>] AS [<c.name>]"
```
This seems bad, to embed details about how we construct a DRS URL inside a select statement. Is there a reason why we don't build the DRS URL after the query has returned?

nmalfroy replied:
I disagree.  This is doing it on the fly and this is the latest place we can put it (since this is what generates the object that is then shared with users directly in azure).  This also is what we do in BQ

pshapiro4broad replied:
OK, I understand. The URLs must be stored as DRS since users access BQ or parquet data directly. The snapshot table in parquet (and view in BQ) are created entirely in SQL so it would be awkward to round trip the data to java just to do the DRS construction there.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java:
```diff
@@ -152,15 +167,38 @@
+        case DATASET:
+          while (resultSet.next()) {
+            String refId = resultSet.getString(refColumn.getName());
+            if (!StringUtils.isEmpty(refId)) {
```
Is this a null check or an empty String check? What is the case where the result is an empty string?

nmalfroy replied:
A case where no data is loaded into a fileref field for a row

pshapiro4broad replied:
So that returns an empty string, not `null`?

nmalfroy replied:
Ah, I was thinking it cold technically be both.  This felt safer without having to know the implementation of how Synapse returns missing fields.  A user could theoretically ingest both

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java:
```diff
@@ -152,15 +167,38 @@
   }
 
+  private Optional<String> extractFileIdFromDrs(final String drsUri) {
+    if (StringUtils.isEmpty(drsUri)) {
```
If this is really a `null` check (vs an empty string check) the whole block can be inside an `Optional`.
```
    return Optional.ofNullable(drsUri).map(URI::create).map(uri -> {
      String fileName = DrsService.getLastNameFromPath(uri.getPath());
      DrsId drsId = drsIdService.fromObjectId(fileName);
      return drsId.getFsObjectId();
  }
```

nmalfroy replied:
Oh, I agree but I didn't think we wanted to use syntax like that

pshapiro4broad replied:
Here it seems fine to me since the method is returning `Optional`. The case I'd rather avoid is where the code doesn't need to return `Optional` but is using it to avoid `if (x != null)`.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java:
```diff
@@ -152,15 +167,38 @@
     }
   }
 
+  private Optional<String> extractFileIdFromDrs(final String drsUri) {
```
I'm hoping this isn't here because of the SELECT above that constructs the DRS URL.. if so it seems like we're making things harder for ourselves by creating a string that we later need to parse.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java:
```diff
@@ -171,14 +209,24 @@
+  private final DrsIdService drsIdService;
+
   @Autowired
-  public AzureSynapsePdao(AzureResourceConfiguration azureResourceConfiguration) {
```
Not sure how the code ended up in this form, but the usual order of class members is:
- fields
- constructors
- methods

So I'd expect the methods above this point to occur after the constructor.

nmalfroy replied:
Yeah, that makes sense.  I'll re-org

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/azure/blobstore/AzureBlobStorePdao.java:
```diff
@@ -34,10 +34,10 @@
 import java.util.Set;
 import java.util.UUID;
 import java.util.stream.Stream;
+import org.apache.commons.codec.binary.Hex;
```
Another option, also using a dependency we already have, is `org.springframework.security.crypto.codec.Hex`.

nmalfroy replied:
I mean, we already have this one too

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -621,7 +621,9 @@ private SnapshotModel populateSnapshotModelFromSnapshot(
       snapshotModel.profileId(snapshot.getProfileId());
     }
     if (include.contains(SnapshotRequestAccessIncludeModel.DATA_PROJECT)) {
-      snapshotModel.dataProject(snapshot.getProjectResource().getGoogleProjectId());
+      if (snapshot.getProjectResource() != null) {
+        snapshotModel.dataProject(snapshot.getProjectResource().getGoogleProjectId());
+      }
     }
```
Can combine nested `if` using `&&`
```
    if (include.contains(SnapshotRequestAccessIncludeModel.DATA_PROJECT) &&
        snapshot.getProjectResource() != null) {
        snapshotModel.dataProject(snapshot.getProjectResource().getGoogleProjectId());
    }
```

nmalfroy replied:
I guess I found this a little more readable

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/create/SnapshotCreateFlight.java:
```diff
@@ -103,14 +98,20 @@ public SnapshotCreateFlight(FlightMap inputParameters, Object applicationContext
+
       addStep(
           new CreateSnapshotInitializeProjectStep(
               resourceService, firestoreRegion, sourceDatasets, snapshotName),
```
This code is passing both `sourceDatasets` and `firestoreRegion`, but `firestoreRegion` is created entirely from `sourceDatasets`. If it's only needed by this constructor, maybe it would be better to create `firestoreRegion` inside the constructor instead of having the caller do it.

nmalfroy replied:
Ah great point

nmalfroy replied:
I'm pushing this down to the ResourceService

pshapiro4broad commented on src/test/java/bio/terra/service/filedata/azure/blobstore/AzureBlobStorePdaoTest.java:
```diff
@@ -75,8 +75,8 @@
-  // This is the base64 encoding of FOOBAR
-  private static final String BLOB_CONTENT_MD5_B64 = "Rk9PQkFS";
+  // This is bytes of the FOOBAR string in Hex
+  private static final String BLOB_CONTENT_MD5_B64 = "464f4f424152";
```
If it's hex, the constant shouldn't end in `B64`. Although I don't see the point of hardcoding the encoded value here, why not call the `Hex` encode method in the test to generate the expected result? We aren't writing a unit test of the Hex API itself, only our use of it.

nmalfroy replied:
derp.  you're right

Same as my point above.  I like varying between the test and the prod code to make sure we're actually testing a thing

pshapiro4broad commented on src/test/java/bio/terra/service/dataset/DatasetAzureIntegrationTest.java:
```diff
@@ -454,19 +457,42 @@ public void datasetIngestFileHappyPath() throws Exception {
 
     // Do a Drs lookup
     String drsId = String.format("v1_%s_%s", snapshotId, fileId);
+    assertThat("Expected Drs object Id exists", drsObjectIds.contains(drsId), equalTo(true));
```
`assertThat(x, y, equalTo(true))` is the same as `assertThat(x, y)`.
```suggestion
    assertThat("Expected Drs object Id exists", drsObjectIds.contains(drsId));
```

nmalfroy replied:
ooh, right, I always forget about that one

pshapiro4broad commented on src/test/java/bio/terra/service/dataset/DatasetAzureIntegrationTest.java:
```diff
@@ -454,19 +457,42 @@ public void datasetIngestFileHappyPath() throws Exception {
     String filePath = result.getLoadFileResults().get(0).getTargetPath();
 
     // Do a Drs lookup
     String drsId = String.format("v1_%s_%s", snapshotId, fileId);
```
Whenever I see code building IDs from scratch like this I wonder what we'll do if we ever change the format. The format itself is supposed to be opaque so users shouldn't be depending on it. Our tests should be written the same way and use a (non-test) Util method to build DRS IDs instead of doing it by hand like this does.

nmalfroy replied:
I disagree here too.  Then the test isn't testing anything.  This has the benefit of extra-reminding the developer that that if the format changes, then it's going to break things downstream

pshapiro4broad replied:
Since there's no way to refactor the SQL code anyway there's not much we can do to eliminate the duplication here. But I disagree with the philosophy that backward compatibility should be tested as a side effect of a test of something else. I'd prefer to have a test that's explicitly for compatibility instead, if we think that's an important thing to do.

pshapiro4broad commented on src/test/java/bio/terra/service/dataset/DatasetAzureIntegrationTest.java:
```diff
@@ -454,19 +457,42 @@ public void datasetIngestFileHappyPath() throws Exception {
+        // TODO: once we have an endpoint to expose parquet data, we should use that mechanism here
+        if (table.getName().equals("vocabulary")) {
+          List<Map<String, String>> records = ParquetUtils.readParquetRecords(tableUrl);
+          assertThat("2 rows are present", records.size(), equalTo(2));
```
I think`hasSize()` is more idiomatic
```suggestion
          assertThat("2 rows are present", records, hasSize(2));
```

nmalfroy replied:
oh cool

pshapiro4broad commented on src/test/java/bio/terra/service/dataset/DatasetAzureIntegrationTest.java:
```diff
@@ -454,19 +457,42 @@ public void datasetIngestFileHappyPath() throws Exception {
+          List<Map<String, String>> records = ParquetUtils.readParquetRecords(tableUrl);
+          assertThat("2 rows are present", records.size(), equalTo(2));
+          // Extract the DRS Ids
+          drsIds.addAll(
+              records.stream()
+                  .map(r -> r.get("vocabulary_reference"))
+                  .collect(Collectors.toList()));
```
Alternate form, a bit shorter
```suggestion
          records.stream()
              .map(r -> r.get("vocabulary_reference"))
              .forEach(drsIds::add);
```

nmalfroy replied:
ooh, will do

pshapiro4broad reviewed (approved):
Looks OK as far as I can tell.

snf2ye reviewed (approved):
LGTM!

nmalfroy committed:
Address PR comments:

- Variable renames
- Moved firestore region detection logic down into ResourceService class
- Re-orged methods in AzureSynapsePdao
- Test code cleanup

samanehsan reviewed (approved):
LGTM! The commit-breakdown was really helpful! :octocat:
